contrib/mesos/pkg/scheduler/podtask/procurement.go

```diff
@@ -17,31 +17,53 @@ limitations under the License.
 package podtask
 
 import (
+	"fmt"
+	"math"
+
 	log "github.com/golang/glog"
 	mesos "github.com/mesos/mesos-go/mesosproto"
+	"github.com/mesos/mesos-go/mesosutil"
+	"k8s.io/kubernetes/contrib/mesos/pkg/scheduler/executorinfo"
 	mresource "k8s.io/kubernetes/contrib/mesos/pkg/scheduler/resource"
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/labels"
 )
 
 // NewDefaultProcurement returns the default procurement strategy that combines validation
 // and responsible Mesos resource procurement. c and m are resource quantities written into
 // k8s api.Pod.Spec's that don't declare resources (all containers in k8s-mesos require cpu
 // and memory limits).
-func NewDefaultProcurement(c mresource.CPUShares, m mresource.MegaBytes) Procurement {
-	resourceProcurer := &RequirePodResources{
-		defaultContainerCPULimit: c,
-		defaultContainerMemLimit: m,
-	}
+func NewDefaultProcurement(prototype *mesos.ExecutorInfo, eir executorinfo.Registry) Procurement {
 	return AllOrNothingProcurement([]Procurement{
-		ValidateProcurement,
-		NodeProcurement,
-		resourceProcurer.Procure,
-		PortsProcurement,
-	}).Procure
+		NewNodeProcurement(),
+		NewPodResourcesProcurement(),
+		NewPortsProcurement(),
+		NewExecutorResourceProcurer(prototype.GetResources(), eir),
+	})
+}
+
+// The ProcurementFunc type is an adapter to use ordinary functions as Procurement implementations.
+type ProcurementFunc func(*T, *mesos.Offer, *api.Node, *Spec) error
+
+func (p ProcurementFunc) Procure(t *T, o *mesos.Offer, n *api.Node, s *Spec) error {
+	return p(t, o, n, s)
 }
 
-// Procurement funcs allocate resources for a task from an offer.
-// Both the task and/or offer may be modified.
-type Procurement func(*T, *mesos.Offer) error
+// Procurement is the interface that implements resource procurement.
+//
+// Procure procurs offered resources for a given pod task T
+// and stores the procurement result in the given Spec.
+// It returns an error if the procurement failed.
+//
+// Note that the T struct also includes a Spec field.
+// This differs from the given Spec parameter which is meant to be filled
+// by a chain of Procure invocations (procurement pipeline).
+//
+// In contrast T.Spec is meant not to be filled by the procurement chain
+// but rather by a final scheduler instance.
```

Comment:
should also note that `mesos.Offer.Resources` may be modified by a procurement func